### PR TITLE
feature: Add support for native PT DDP distribution

### PR DESCRIPTION
**Pull request by vishwakaria**
*Issue #, if available:*

*Description of changes:*
* Allow customers to launch native PTDDP jobs with `nccl` backend using a new distribution value: `{ "pytorchddp": "enabled": "true",  <other options> }}`
* Uses smdataparallel runner for launch, same as smdistributed backend.
* [Design document](https://quip-amazon.com/XqmTAXW29Nua/Improving-Launcher-experience-on-SageMaker)
* Connected PRs: https://github.com/aws/sagemaker-pytorch-training-toolkit/pull/236

*Testing done:*
* https://quip-amazon.com/usIaAMvBbknO/Test-Plan-PT-DDP-Launcher

*Note: This PR can only be merged after the sagemaker-pytorch-training-toolkit PR has been merged.* 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ X ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ X ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ N/A ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ X ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ X ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ X ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ N/A ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by navinsoni — APPROVED** [review]

**Review by BasilBeirouti — APPROVED** [review]

File: src/sagemaker/pytorch/estimator.py [hunk 1]
```diff
@@ -199,6 +214,9 @@ def __init__(
             if self.framework_version and Version(self.framework_version) >= Version("1.3"):
                 kwargs["enable_sagemaker_metrics"] = True
 
+        if "instance_type" in kwargs:
+            self.instance_type = kwargs["instance_type"]
```
Comment (yl-to) [inline src/sagemaker/pytorch/estimator.py:218]: is `self.instance` already be assigned in the base class's init func?

File: src/sagemaker/pytorch/estimator.py [hunk 2]
```diff
@@ -215,10 +233,28 @@ def __init__(
...
+        """Returns a dict of distribution config for PyTorch training
+
+        Args:
+            distribution (dict): A dictionary with information on how to run distributed training.
+        Returns:
+            dict containing Pytorch DDP config
+        """
+        distribution_config = {}
+        if "pytorchddp" in distribution:
+            pytorch_ddp_enabled = distribution.get("pytorchddp").get("enabled", False)
```
Comment (yl-to) [inline src/sagemaker/pytorch/estimator.py:246]: What if the user set up `pytorchddp` as `False`, and set normal distribution config. looks these lines didn't work for this case.

Comment (vishwakaria) [reply src/sagemaker/pytorch/estimator.py:246]: Makes sense, fixed.

File: src/sagemaker/pytorch/estimator.py (hunk 1 above)
Comment (vishwakaria) [reply src/sagemaker/pytorch/estimator.py]: Yes, removed.

**Review by yl-to — APPROVED** [review]
LGTM

File: src/sagemaker/fw_utils.py [hunk 3]
```diff
@@ -103,6 +103,17 @@
         "1.11.0",
     ],
 }
+
+PYTORCHDDP_SUPPORTED_FRAMEWORK_VERSIONS = [
+    "1.10",
```
Comment (roywei) [inline src/sagemaker/fw_utils.py:108]: I don't see 1.10.2 here. Also are these versions essentially use the same DLC docker? which is 1.10.2? as we only have one DLC docker file for each minor release.

File: src/sagemaker/fw_utils.py [hunk 4]
```diff
@@ -747,12 +764,56 @@ def validate_distribution(
...
+        ValueError: if
+            `py_version` is not python3 or
+            `framework_version` is not in PYTORCHDDP_SUPPORTED_FRAMEWORK_VERSIONS
+    """
+    if framework_name != "pytorch":
+        # We need to validate only for PyTorch framework
+        return
+
+    err_msg = ""
+    if not image_uri:
```
Comment (roywei) [inline src/sagemaker/fw_utils.py:798]: what will happen if using an custom docker image uri and there is old toolkit installed there?

File: tests/conftest.py [hunk 5]
```diff
@@ -411,6 +411,18 @@ def tf_full_py_version(tf_full_version):
     return "py39"
 
 
+@pytest.fixture(scope="module")
+def pytorch_ddp_py_version():
+    return "py3"
+
+
+@pytest.fixture(
+    scope="module", params=["1.10", "1.10.0", "1.10.1", "1.11", "1.11.0", "1.12", "1.12.0"]
```
Comment (roywei) [inline tests/conftest.py:420]: 1.10.2 is missing, also is 1.10.1 a valid version that we released? we may have skipped it

**Review by roywei — COMMENTED** [review]
overall LGTM, few comments.

File: src/sagemaker/fw_utils.py (hunk 4 above)
Comment (vishwakaria) [reply src/sagemaker/fw_utils.py:798]: I am mirroring the logic that we have for smdistributed: https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/fw_utils.py#L635.
How can we check for older toolkit versions inside custom DLCs?

File: tests/conftest.py (hunk 5 above)
Comment (vishwakaria) [reply tests/conftest.py:420]: Gotcha, I will switch to 1.10.2 then.

File: src/sagemaker/fw_utils.py (hunk 3 above)
Comment (vishwakaria) [reply src/sagemaker/fw_utils.py:108]: Adding 1.10.2. 
Yeah, seems like there is only one DLC docker file for 1.10 (which is for 1.10.2): https://github.com/aws/deep-learning-containers/blob/master/available_images.md#prior-sagemaker-framework-container-versions

**Comment by vishwakaria** [comment]
Moved to https://github.com/aws/sagemaker-python-sdk/pull/3270 after rebase and squashing commits.

**Pull request closed without merging**